wire-ios-data-model/Source/E2EIdentity/CRL/CertificateRevocationListsChecker.swift

```diff
@@ -0,0 +1,138 @@
+        context: NSManagedObjectContext
+    ) {
+        self.init(
+            crlAPI: crlAPI,
+            crlExpirationDatesRepository: CRLExpirationDatesRepository(userID: userID),
+            mlsConversationsVerificationUpdater: mlsConversationsVerificationUpdater,
+            coreCryptoProvider: coreCryptoProvider,
+            context: context
+        )
+    }
+    
+    init(
+        crlAPI: CertificateRevocationListAPIProtocol,
+        crlExpirationDatesRepository: CRLExpirationDatesRepositoryProtocol,
+        mlsConversationsVerificationUpdater: MLSConversationVerificationStatusUpdating,
+        coreCryptoProvider: CoreCryptoProviderProtocol,
+        context: NSManagedObjectContext
+    ) {
+        self.crlAPI = crlAPI
+        self.crlExpirationDatesRepository = crlExpirationDatesRepository
+        self.mlsConversationsVerificationUpdater = mlsConversationsVerificationUpdater
+        self.coreCryptoProvider = coreCryptoProvider
+        self.context = context
+    }
+
+    public func checkNewCRLs(from distributionPoints: CRLsDistributionPoints) async {
+
+        let newDistributionPoints = distributionPoints.urls.filter {
+            !crlExpirationDatesRepository.crlExpirationDateExists(for: $0)
+        }
+
+        await checkCertificateRevocationLists(from: newDistributionPoints)
+    }
+
+    public func checkExpiringCRLs() async {
+        let distributionPointsOfExpiringCRLs = crlExpirationDatesRepository
+            .fetchAllCRLExpirationDates()
+            .filter({ isCRLExpiringSoon(expirationDate: $0.value) })
+            .keys
+
+        await checkCertificateRevocationLists(from: Set(distributionPointsOfExpiringCRLs))
+    }
+
+    private func checkCertificateRevocationLists(from distributionPoints: Set<URL>) async {
+
+        for distributionPoint in distributionPoints {
+            do {
+                // fetch the CRL from the distribution point
+                let crlData = try await crlAPI.getRevocationList(from: distributionPoint)
+
+                // register the CRL with core crypto
+                let registration = try await coreCrypto.perform {
+                    try await $0.e2eiRegisterCrl(crlDp: distributionPoint.absoluteString, crlDer: crlData)
+                }
+
+                // store the expiration time
+                if let expirationTimestamp = registration.expiration {
+                    let expirationDate = Date(timeIntervalSince1970: TimeInterval(expirationTimestamp))
+                    crlExpirationDatesRepository.storeCRLExpirationDate(expirationDate, for: distributionPoint)
+                }
+
+                // check if certificate is "dirty"
+                if registration.dirty {
+                    // update verification state for conversations
+                    try await mlsConversationsVerificationUpdater.updateAllStatuses()
+                }
+            } catch {
+                // TODO: Log
```

Comment:
> should we maybe bubble the first thrown error?

I guess it depends if we want to abort the whole operation or not, since we iterate through all the distribution points